### PR TITLE
Trigger autorecovery against clustered rabbitmq instances 

### PR DESCRIPTION
**Pull request by oscarfh**
This aims to fix https://github.com/micronaut-projects/micronaut-rabbitmq/issues/210
My goal was to have a recovery of the channel only, but this is not possible with the data I have available in the Advice code, so I had to force a full (connection, bindings, topology etc) reconnection. I was also not able to use the retry mechanism that is already implemented in the AutorecoveryConnection class (by means of the listener) because a fail in the topology caused by a missing queue due to the missing node causes the recovery process to hang.
Therefore I had to retry the connection and consumption myself until it works. Fortunately I was able to use the delay configured by the user.

I also tried to extract the anonymous DefaultConsumer to its own class, unfortunately this is not easily doable due to the integration between the anonymous class and its parent class.

### CODE REVIEW

**Comment by oscarfh** [comment]
It would be very nice if anyone could help me with the flaky build.

File: rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java [hunk 1]
```diff
@@ -75,26 +79,30 @@
...
+     * @param conversionService       The service to convert consume argument values
+     * @param connectionFactoryConfig The configuration of the rabbit connection
      */
     public RabbitMQConsumerAdvice(BeanContext beanContext,
                                   RabbitBinderRegistry binderRegistry,
                                   RabbitListenerExceptionHandler exceptionHandler,
                                   RabbitMessageSerDesRegistry serDesRegistry,
-                                  ConversionService conversionService) {
+                                  ConversionService conversionService,
+                                  RabbitConnectionFactoryConfig connectionFactoryConfig) {
```
Comment (jameskleeh) [inline rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java:99]: This is a public class not marked with `@Internal` so this is a breaking change. A new constructor can be created and marked with `@Inject` and you'll have to deal with the parameter potentially being null

File: rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java [hunk 2]
```diff
@@ -193,111 +201,213 @@ public void process(BeanDefinition<?> beanDefinition, ExecutableMethod<?, ?> met
...
+            @Override
+            public void handleTerminate(String consumerTag) {
+                if (channel instanceof RecoverableChannel) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug(
+                                "The channel was terminated.  Automatic recovery attempt is underway for consumer [{}]",
+                                consumerParams.clientTag);
+                    }
+                    try {
+                        channel.basicConsume(consumerParams.queue, false, consumerParams.clientTag, false, consumerParams.exclusive, arguments, this);
```
Comment (jameskleeh) [inline rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java:246]: So in general it seems this PR is working around or trying to amend the functionality built into the java client API. How can you safely call consume here when its possible for RabbitMQ to auto recover? Couldn't this end up in multiple subscriptions?

**Comment by jameskleeh** [comment]
I think it would be good if we can simulate this somehow in a test. Perhaps that means shutting down the testcontainers instance and restarting it. Without some sort of test I can't know if it actually works. Whatever you are doing to assert that these changes are working should be doable in a test.

**Comment by oscarfh** [comment]
@jameskleeh  The problem is that we need a clustered RabbitMQ for that. That is the only reason I did not make a test

**Comment by oscarfh** [comment]
I will close this without merging because I cold confirm that this generates multiple subscriptions.

**Comment by jameskleeh** [comment]
@oscarfh Sorry it didn't work out. I would recommend looking at the RabbitMQ Java driver to see if there are any changes you can make there

File: rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java (hunk 2 above)
Comment (michaelklishin) [reply rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java:246]: When RabbitMQ closes a channel, it is [due to an error](https://www.rabbitmq.com/channels.html#error-handling). RabbitMQ Java client explicitly [does not recover](https://www.rabbitmq.com/api-guide.html#recovery) from such failures. IMO no client should. This is a clear case where the error should be visible to the user.

**Comment by michaelklishin** [comment]
If the recovery is tripped up by a missing queue because it is [not replicated](https://www.rabbitmq.com/ha.html#non-mirrored-queue-behavior-on-node-failure) and its home node is unavailable, the solution is not to fix Micronaut but to use a [modern replicated queue type](https://www.rabbitmq.com/quorum-queues.html). By using a non-replicated queue type the user explicitly agrees to tie the availability of said queue to a single node.

We would consider changes to the Java client that would make topology recovery retries more useful here but note that for as long as the node hosting the non-replicated queue is down, the recovery of such client would be blocked on retries, which again, sounds like something that should not be done and clients should not try to recover from. Instead, a replicated queue type (a quorum queue) should be used.

**Pull request closed without merging**